contracts/dynamic-contracts/extension/PermissionsSig.sol

```diff
@@ -0,0 +1,310 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.0;
+
+/// @author thirdweb
+
+import "../../extension/interface/IPermissionsSig.sol";
+import "../../lib/TWStrings.sol";
+import "../../openzeppelin-presets/utils/cryptography/EIP712.sol";
+
+/**
+ *  @title   Permissions
+ *  @dev     This contracts provides extending-contracts with role-based access control mechanisms
+ */
+
+library PermissionsSigStorage {
+    bytes32 public constant PERMISSIONS_STORAGE_POSITION = keccak256("permissions.sig.storage");
+
+    struct Data {
+        /// @dev Map from keccak256 hash of a role => a map from address => whether address has role.
+        mapping(bytes32 => mapping(address => bool)) _hasRole;
+        /// @dev Map from keccak256 hash of a role to role admin. See {getRoleAdmin}.
+        mapping(bytes32 => bytes32) _getRoleAdmin;
+        /// @dev Mapping from a signed request UID => whether the request is processed.
+        mapping(bytes32 => bool) executed;
+    }
+
+    function permissionsSigStorage() internal pure returns (Data storage permissionsData) {
+        bytes32 position = PERMISSIONS_STORAGE_POSITION;
+        assembly {
+            permissionsData.slot := position
+        }
+    }
+}
+
+abstract contract PermissionsSigUtils is IPermissionsSig, EIP712 {
+    using ECDSA for bytes32;
+
+    bytes32 private constant TYPEHASH =
+        keccak256(
+            "RoleRequest(bytes32 role,address target,uint8 action,uint128 validityStartTimestamp,uint128 validityEndTimestamp,bytes32 uid)"
+        );
+
+    constructor() EIP712("PermissionsSig", "1") {}
+
+    /*///////////////////////////////////////////////////////////////
+                            External functions
+    //////////////////////////////////////////////////////////////*/
+
+    /// @dev Verifies that a request is signed by an authorized account.
+    function verifyRoleRequest(RoleRequest calldata req, bytes calldata signature)
+        public
+        view
+        virtual
+        returns (bool success, address signer)
+    {
+        PermissionsSigStorage.Data storage data = PermissionsSigStorage.permissionsSigStorage();
+        signer = _recoverAddress(req, signature);
+        success = !data.executed[req.uid] && _isAuthorizedSigner(signer, req, data);
```

Comment:
so here its not `isAuthorizedSigner` but rather `isAdmin` of `canModifyRoles` right?

Reply:
I kept the naming general i.e. `isAuthorizedSigner` because the criteria for being a valid signer changes based on grant/revokeRole and renounceRole. 

For grant and revoke Role functions: the signer must have the role that is the role admin of the role being granted / revoked. For renounceRole: the signer must be the account renouncing the role.

This is the behaviour of regular `Permissions`. I tried not to change this core behaviour.